okta-saml/main.py

```
@@ -203,5 +254,40 @@ def metadata():
+    port = int(os.environ.get('PORT', '8443'))
+
+    # Auto-fetch remote metadata if configured and local file missing
+    try:
+        if METADATA_URL and not os.path.exists(METADATA_LOCAL):
+            try:
+                # Use stdlib to avoid external dependency on `requests`
+                from urllib.request import urlopen, Request
+                from urllib.error import URLError, HTTPError
+
+                print(f"Fetching metadata from: {METADATA_URL}")
+                req = Request(METADATA_URL, headers={"User-Agent": "metadata-fetcher/1.0"})
```

Ruff (0.14.11)

Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)